lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1622,11 +1622,11 @@ bool TypeChecker::coerceParameterListToType(ParameterList *P, ClosureExpr *CE,
         param->setInterfaceType(CE->mapTypeOutOfContext(ty));
       }
     }
-
-    if (!ty->isMaterializable()) {
-      if (ty->is<InOutType>()) {
-        param->setLet(false);
-      } else if (param->hasName()) {
+    
+    if (ty->is<InOutType>() || ty->isLValueType()) {
```

Comment:
But I don't think the LValueType case can. Try adding an assert and running the source compat suite.